lib/posix-event/eventfd.c

```diff
@@ -0,0 +1,423 @@
+		events |= EPOLLOUT;
+	if (efd->val == (uint64_t)-1)
+		events |= EPOLLERR;
+
+	return events;
+}
+
+static void eventfd_signal_eventpoll(struct eventfd *efd, unsigned int events)
+{
+	struct eventpoll_cb *ecb;
+	struct uk_list_head *itr;
+
+	uk_list_for_each(itr, &efd->ep_list) {
+		ecb = uk_list_entry(itr, struct eventpoll_cb, cb_link);
+
+		UK_ASSERT(ecb->unregister);
+
+		eventpoll_signal(ecb, events);
+	}
+}
+
+static int eventfd_vfscore_close(struct vnode *vnode,
+				 struct vfscore_file *fp __unused)
+{
+	struct eventfd *efd;
+
+	UK_ASSERT(vnode->v_data);
+	UK_ASSERT(vnode->v_type == VEVENT);
+
+	efd = (struct eventfd *)vnode->v_data;
+
+	uk_free(uk_alloc_get_default(), efd);
+
+	vnode->v_data = NULL;
+	return 0;
+}
+
+static int eventfd_vfscore_read(struct vnode *vnode,
+				struct vfscore_file *fp,
+				struct uio *buf, int ioflag __unused)
+{
+	struct eventfd *efd = (struct eventfd *)vnode->v_data;
+	uint64_t *val;
+
+	UK_ASSERT(vnode->v_data);
+	UK_ASSERT(vnode->v_type == VEVENT);
+
+	if (unlikely(buf->uio_offset != 0))
+		return EINVAL;
+
+	if (unlikely(buf->uio_iovcnt != 1))
+		return EINVAL;
+
+	if (unlikely(!buf->uio_iov[0].iov_base))
+		return EINVAL;
+
+	if (unlikely(buf->uio_iov[0].iov_len < sizeof(uint64_t)))
+		return EINVAL;
+
+	val = (uint64_t *)buf->uio_iov[0].iov_base;
+
+	uk_mutex_lock(&efd->lock);
+
+	/* Check if the value is 0. In this case, block if the file descriptor
+	 * is not configured to non-blocking operation.
+	 */
+	if ((efd->val == 0) && (fp->f_flags & O_NONBLOCK)) {
+		uk_mutex_unlock(&efd->lock);
+		return EAGAIN;
+	}
+
+	uk_waitq_wait_event_mutex(&efd->r_wq, efd->val > 0, &efd->lock);
+
+	UK_ASSERT(efd->val > 0);
+
+	*val = (efd->is_semaphore) ? 1 : efd->val;
+	efd->val -= *val;
+
+	eventfd_signal_eventpoll(efd, EPOLLOUT);
+
+	uk_waitq_wake_up(&efd->w_wq);
+
+	uk_mutex_unlock(&efd->lock);
+
+	return 0;
+}
+
+static int eventfd_vfscore_write(struct vnode *vnode, struct uio *buf,
+				 int ioflag __unused)
+{
+	struct eventfd *efd = (struct eventfd *)vnode->v_data;
+	uint64_t val;
+
+	UK_ASSERT(vnode->v_data);
+	UK_ASSERT(vnode->v_type == VEVENT);
+
+	if (unlikely(buf->uio_offset != 0))
+		return EINVAL;
+
+	if (unlikely(buf->uio_iovcnt != 1))
+		return EINVAL;
+
+	if (unlikely(!buf->uio_iov[0].iov_base))
+		return EINVAL;
+
+	if (unlikely(buf->uio_iov[0].iov_len != sizeof(uint64_t)))
+		return EINVAL;
+
+	val = *((uint64_t *)buf->uio_iov[0].iov_base);
+	if (unlikely(val == (uint64_t)-1))
+		return EINVAL;
+
+	uk_mutex_lock(&efd->lock);
+
+	/* Check if the addition would overflow. In this case, block if the
+	 * file descriptor is not configured to non-blocking operation.
+	 *
+	 * TODO: Make fp available here
+	 */
+	/*
+	if ((efd->val >= (uint64_t)-1 - val) && (fp->f_flags & O_NONBLOCK)) {
+		uk_mutex_unlock(&efd->lock);
+		return EAGAIN;
+	}
+	*/
+
+	uk_waitq_wait_event_mutex(&efd->w_wq, efd->val < (uint64_t)-1 - val,
+				  &efd->lock);
+
+	UK_ASSERT(efd->val < (uint64_t)-1 - val);
```

Comment:
I get this, still, it feels a bit wrong. Maybe there is a macro somewhere for this already.

Reply:
Introduced the following macro:
```C
#define eventfd_no_overflow(cur, val)				\
	((cur) < UINT64_MAX - (val))
```

Reply:
Perfect